ipa-macros/src/derive_gate/mod.rs

```diff
@@ -0,0 +1,392 @@
+//         2) annotating the steps that are conditionally executed. We need to take into consideration about
+//         a case where there are child steps branching off from the conditional step.
+//     (b) Forcibly execute conditional steps when `step-trace` feature is on to generate the steps file.
+//         Not sure what is the best way to do this, or whether this will work though.
+//  - Currently, we are doing (a-1). It could let unexpected state transitions happen without panicking, but
+//    there isn't a better way to solve this.
+//
+//   Examples:
+//   - crate::protocol::boolean::random_bits_generator::FallbackStep
+//     This step is only executed when generated random secret shared bits are of a share larger than
+//     `2^32 - 5`. However, the context is narrowed to this step when a RBG instance is created. This one
+//     is easy to solve by allowing to narrow to a hard-coded state as there are no branching child steps.
+//   - crate::protocol::attribution::Step::CurrentCreditOrCreditUpdate
+//     This and other similar steps in the "binary tree prefixed sum" protocols are highly optimized such
+//     that they are executed (or not executed) depending on the iteration/depth of the loop. To solve this,
+//     we could apply the same conditions used to call `multiply` using these contexts to where we `narrow`
+//     the context. The code becomes a bit messy, but it's doable.
+//   - crate::protocol::context::UpgradeStep
+//     This steps is executed in both malicious and semi-honest contexts, but the `narrow` call in semi-honest
+//     context is a dummy; it doesn't trigger `send`. This is a bit tricky to solve because 1) there are many
+//     child steps, and 2) malicious context
+//
+// 4. (Done) Generate state transitions of dynamic steps. (i.e., BitOpStep)
+//   - There are steps that are dynamically generated based on the number of bits, rows, etc. Fortunately,
+//     these steps have finite number of states, so we can generate the state transitions either in steps
+//     file generation or in compile time. Again, we need to take into consideration about the case where
+//     there are child steps branching off from the dynamic step.
+//   - Currently we do this in `collect_steps.py`. In the future, we want to do this in compile time with
+//     a proc macro, but that requires a feature currently in nightly (mentioned in #1).
+//
+// 5. num-multi-bits
+//   - `num-multi-bits` also changes the state transition map. We could generate the steps file for all
+//     possible values of `num-multi-bits, but that will make the file huge. We can probably just stick to
+//     the current empirical best value of 3. We could also try to read the value from the source or the
+//     config file, and generate the steps file accordingly. However, the value could change after the steps
+//     file is generated, so we need to make sure that the steps file is always up to date somehow.
+//
+// 6. Root step
```

Comment:
We might need a separate type and proc macro for that.

Reply:
This was more about whether we need to dynamically generate a string representation of root step(s), i.e., does query ID "X" must be generated by `Compact` gate's `AsRef<str>` implementation? If we do not require the root step in a real world scenario, we don't need to worry about this. `Descriptive` gate has `id` field which embeds `protocol/run-0` as the root step, so I was wondering if that was also needed for `Compact` gate.